src/index.js

```diff
@@ -185,7 +185,115 @@ const getEncodingIdx = (type) => {
+ *   colorBy: DataEncoding;
+ *   deselectOnDblClick: boolean;
+ *   deselectOnEscape: boolean;
+ *   lassoColor: Color | string;
+ *   lassoLineWidth: number;
+ *   lassoMinDelay: number;
+ *   lassoMinDist: number;
+ *   lassoClearEvent: LassoClearEvent;
+ *   lassoInitiator: boolean;
+ *   keyMap: Record<Key, Omit<Action, 'panZoom'>>;
+ *   mouseMode: Omit<Action, 'merge'>;
+ *   showReticle: boolean;
+ *   reticleColor: Color | string;
+ *   pointColor: Color;
+ *   pointColorActive: Color;
+ *   pointColorHover: Color;
+ *   showPointConnections: boolean;
+ *   pointConnectionColor: Color;
+ *   pointConnectionColorActive: Color;
+ *   pointConnectionColorHover: Color;
+ *   pointConnectionColorBy: PointDataEncoding;
+ *   pointConnectionOpacity: number;
+ *   pointConnectionOpacityBy: PointDataEncoding;
+ *   pointConnectionOpacityActive: number;
+ *   pointConnectionSize: number;
+ *   pointConnectionSizeActive: number;
+ *   pointConnectionSizeBy: PointDataEncoding;
+ *   pointConnectionMaxIntPointsPerSegment: number;
+ *   pointConnectionTolerance: number;
+ *   pointSize: number;
+ *   pointSizeSelected: number;
+ *   pointSizeMouseDetection: 'auto';
+ *   pointOutlineWidth: number;
+ *   opacity: number;
+ *   opacityBy: DataEncoding;
+ *   opacityByDensityFill: number;
+ *   sizeBy: DataEncoding;
+ *   height: 'auto' | number;
+ *   width: 'auto' | number;
+ *   gamma: number;
+ * }} Settable
+ *
+ * @typedef {{
+ *   canvas: HTMLElement;
+ *   regl: import('regl').Regl;
+ *   syncEvents: boolean;
+ *   version: string;
+ *   lassoInitiatorElement: HTMLElement;
+ *   camera: Camera2D;
+ *   performanceMode: boolean;
+ *   opacityByDensityDebounceTime: number;
+ * } & Settable} Properties
+ *
+ * @param {Partial<Properties>} initialProperties
+ */
 const createScatterplot = (initialProperties = {}) => {
+  /**
+   * @typedef {{
+   *  camera: Camera2D;
+   *  view: any;
+   *  xScale: [number, number];
+   *  yScale: [number, number];
+   * }} DrawPayload
+   *
+   * @typedef {{
+   *   init: () => void;
+   *   destroy: () => void;
+   *   backgroundImageReady: () => void;
+   *   pointOver: (pointIndex: number) => void;
+   *   pointOut: (pointIndex: number) => void;
+   *   select: (payload: { points: number[][] }) => void;
+   *   deselect: () => void;
+   *   view: (payload: DrawPayload) => void;
+   *   draw: (payload: DrawPayload) => void;
+   *   lassoStart: () => void;
+   *   lassoExtend: (payload: { coordinates: number[] }) => void;
+   *   lassoEnd: (payload: { coordinates: number[] }) => void;
+   *   transtionStart: () => void;
+   *   transitionEnd: (payload: import('regl').Regl) => void;
+   *   pointConnectionsDraw: () => void;
+   * }} PubSubEvents
+   *
+   * @type {{
+   *  subscribe: <Event extends keyof PubSubEvents>(event: Event, handler: PubSubEvents[Event], times: number) => void;
+   *  unsubscribe: <Event extends keyof PubSubEvents>(event: Event) => void;
+   * }}
+   */
```

Comment:
I'm rather happy with this bit. Based on the event name, TS will infer the signature of the callback:

<img width="801" alt="image" src="https://user-images.githubusercontent.com/24403730/123800796-a68ff880-d8b7-11eb-8a66-37d6377fd350.png">